### ISSUE TITLE
Enhance "Jsonb" to allow "fromJson" to handle "JsonStructure" as input

### ISSUE THREAD

**Issue by michael-schnell** [eclipse-ee4j/jsonb-api#111]
Hey there,

It would be nice to enhance the interface [javax.json.bind.Jsonb](https://github.com/eclipse-ee4j/jsonb-api/blob/master/api/src/main/java/javax/json/bind/Jsonb.java) to allow handling of [javax.json.JsonStructure](https://github.com/eclipse-ee4j/jsonp/blob/master/api/src/main/java/javax/json/JsonStructure.java) as input for deserialization (unmarshalling to a Java class).

This means, there should be two new methods :
```java
public interface Jsonb extends AutoCloseable {
  ...
  <T> T fromJson(JsonStructure jsonStruct, Class<T> type) throws JsonbException;
  <T> T fromJson(JsonStructure jsonStruct, Type runtimeType) throws JsonbException;
}
```
Yasson already supports this for the opposite direction (serialization):
```java
JsonObject jsonObj = Json.createObjectBuilder().add("name", "Peter Parker").add("age", 18).build();
Jsonb jsonb = JsonbBuilder.create();
System.out.println(jsonb.toJson(jsonObj));
// Prints {"name":"Peter Parker","age":18}
```
But currently JSON-B does not support deserialization like:
```java
Person person = jsonb.fromJson(jsonObj, Person.class);
```
Unfortunately this is a breaking change, but it could be implemented in the interface with a default method that throws an "UnsupportedOperationException" in case the provider does not implement it.

What do you think?

Cheers,
Michael

**Comment by rmannibucau** [comment]
-1 for a default method throwing an exception, it can be implemented calling toString which returns json for structures

**Comment by michael-schnell** [comment]
I know it's not nice, but your suggestion will just fail some milliseconds later when a String is assigned:
`Person person = jsonb.fromJson(jsonObj, Person.class);` (String is not assignable to the expected "Person" object...

**Comment by rmannibucau** [comment]
No, this what im using in some apps without perf constraints on that part, fromJson(structure.toString(), Person.class) works. Adds a useless processing/serialization but it is functional.

Java does throw this exception for a few things but at spec level it leads to issues and breaks apps quite easily and users just drop the spec or use weird workarounds where the fix is easy, jsonp 1.1 got a lot of troubles with it during 1.0->1.1 time so better to stay away from it IMHO, in particular cause it is trivial here to not use it.

Side note: default methods break the compat in case users or libs did a java proxy since java doesnt enable to call it properly in that case. Not sure how much we care of that case but when it hits you it hurts from experience and depending java version the workaround is not the same so not nice for consumers yet.

Wonder if we can interact with jsonp to provide some mapping like jdbc has unwrap? Anyone already thought of it? Would be a spi jsonp would load and jsonb would have one impl of it. It would provide JsonStructure.as(ExpectedType.class[, mapperType]) or equivalent enabling the conversion at jsonp level. It is just a shortcut for jsonp but makes the api more fluent (can be used to extract a token for instance during oauth2 auth) so can be worthbit.

**Comment by michael-schnell** [comment]
Hmmm... Actually this means you can't change any of the API interface classes anymore. I think in general it does not cause any trouble as the implementations can implement it later. It´s simply not supported until they did it.  The "proxy" problem is indeed something to think about. I must admit I never tried it to see what happens in this case with CDI or Spring DI.

The "toString()" approach is actually my current workaround. But I dislike that, as it add unnecessary memory and perfomance issues in case the JSON object is not that small.

"Unwrapping" might not be always possible if I think implementations like pull parsers or similar. They might not use JSON-P underneath, but still can implement the binding API.

**Comment by michael-schnell** [comment]
At least with CDI there is no problem with proxy and default methods, The WeldClientProxy works well with the following interface enhancement:
```java
default <T> T fromJson(JsonStructure jsonStruct, Class<T> type) throws JsonbException {
	throw new UnsupportedOperationException("Not implemented (yet) by the provider");
}

default <T> T fromJson(JsonStructure jsonStruct, Type runtimeType) throws JsonbException {
	throw new UnsupportedOperationException("Not implemented (yet) by the provider");
}
```

**Comment by aguibert** [comment]
Hi @michael-schnell, thanks for the suggestion.  Before we go too far into compatibility and whether or not to use a default method, I wanted to question the value this new method would provide.  

This API method would essentially allow us to go from JSON-P objects to POJO, but why even create the JSON-P objects in the first place instead of just going straight to POJO.  

For example, instead of doing this:
```java
JsonObject jsonObj = Json.createObjectBuilder().add("name", "Peter Parker").add("age", 18).build();
Person person = jsonb.fromJson(jsonObj, Person.class);
```

Why not just create the `Person` object directly without JSON-B?
```java
Person p = new Person("Peter Parker", 18);`
```

**Comment by michael-schnell** [comment]
Unfortunately JSON-B (binding) has some limitations compared to JSON-B (XML binding). For example there is no support for the "@XmlAnyElement" concept. Therefore you will sooner or later come to a point where you need to implement things like a JsonbSerializer/JsonbDeserializer. 

Let's now assume that you need to be compatible with an existing JSON-B structure. You are not allowed to change it and there is an "@XmlAnyElement" included. This means it's an object you don't know anything about at the point where the [JsonbDeserializer](https://static.javadoc.io/javax.json.bind/javax.json.bind-api/1.0/javax/json/bind/serializer/JsonbDeserializer.html) is processing the event for that key.

Here is a concrete example of how this could look like:
```java
public static class JsonbDeSer implements JsonbDeserializer<ObjWithAnyElement>, ... {

	@Override
	public ObjWithAnyElement deserialize(JsonParser parser, DeserializationContext ctx, Type rtType) {
		ObjWithAnyElement objWithAnyElement = new ObjWithAnyElement();
		JsonObject content = null;
		while (parser.hasNext()) {
			JsonParser.Event event = parser.next();
			if (event == JsonParser.Event.KEY_NAME) {
				String field = parser.getString();
				switch (field) {
					 ... other fields ...
				case "meta":
					objWithAnyElement.meta = ctx.deserialize(Meta.class, parser));
					break;
				case "data":
					content = ctx.deserialize(JsonObject.class, parser);
					break;
				default:
					// ignore
					break;
				}
			}
		}

		// Handle data at the end, because meta data is only safely available at the end of the process
		// Unfortunately we also need to "peek" into the content to see if the content is base64 encoded.
		// This information is NOT available as meta data in this scenario.
		if (content.containsKey("Base64")) {
			// In case there is a "Base64" element, simply construct a base64 container
			objWithAnyElement.data = new Base64Data(content.getString("Base64"));
		} else {
			// All other cases use "meta" information to deserialize the JsonObject
			objWithAnyElement.data = jsonb.fromJson(content, meta.getType());
		}

		return objWithAnyElement;
	}

}
```

In case the field is "data", you deserialize to a JsonObject and reuse that at the end to decide wether it's base64 data or use the meta data to deserialize to any other (known) type.

**Comment by bravehorsie** [comment]
Almost identical as #67
I have not been a fan of adding native support for JsonStructure <-> Pojo conversion, but it looks some find it useful.

@michael-schnell alternatively to deserializing `content` field into JsonObject it would work to move parser till end of the `content` json structure definition, saving to StringBuffer, which could be than deserialized with JSONB into proper type according to metadata. That would not be one-liner using JSONB but would avoid some of a useless processing/serialization.

**Comment by michael-schnell** [comment]
Is it already possible to interact with the parser like described to save the content field into a string buffer or would this require also a change?

**Comment by bravehorsie** [comment]
By such interaction I mean just moving parser by hand until the end of structure appending to String or JsonGenerator. [Here is an example](https://github.com/bravehorsie/yasson-client/blob/master/src/test/java/info/grigoriadi/ReadJsonStructureToStringTest.java) of what I thought.

**Comment by michael-schnell** [comment]
OK, got what you suggested. 

Maybe we could avoid changing the interface by doing it with a little trick similar to what you suggested:

1. Define some kind of interface in the API that signals "Just return me a string for the current structure" which means to return the current field as string or read until "END_ARRAY" or "END_OBJECT".
   ```
   public interface StringContainer {
       public String getString();
   }
   ```
2. Use the class in the "deserialize" method as parameter
    ```
   StringContainer container = ctx.deserialize(StringContainer.class, parser);
   String json = container.getString();
   ```
3. Use JSON-B to deserialize the content
    `objWithAnyElement.data = jsonb.fromJson(json, meta.getType());`

Then it would be possible for the implementation to simply parse until end of the field/structure and store it inside the `StringContainer` implementing class.

**Comment by mdzaebel** [comment]
In my issue (https://github.com/eclipse-ee4j/yasson/issues/259) I supported the requirement, to be able to bind e.g. a JsonObject to it's class instance (not only Streams or Strings). The reason is, that we need to transform the JsonObject in advance and don't need/want an intermediate String representation again to bind the object.

**Comment by kalgon** [comment]
Why limit the enhancement to `JsonStructure` and not `JsonValue`? With `JsonValue`, you could also convert small value objects. This could be useful if you need to store any type of objects (small or complex) in a generic json way.
```java
JsonValue value = Json.createValue("2001-02-03"); // '2001-02-03'
LocalDate date = jsonb.fromJson(value, LocalDate.class);
```

**Comment by michael-schnell** [comment]
Agree. As JsonValue is the top level structure it does make sense to allow using it.

**Comment by rmannibucau** [comment]
+1 (without the default which is not needed for a jakarta spec)

**Comment by michael-schnell** [comment]
Without the default we would break all existing implementations when changing the interface.

**Comment by rmannibucau** [comment]
@michael-schnell this is a false belief in ee land, it would break implementors which is intended and the case anyway. It does not affect users at all. This is why it should never be used in ee.

**Comment by michael-schnell** [comment]
If a breaking change is no problem, I can live without the default.

**Comment by aguibert** [comment]
FYI this was implemented in Yasson API in this PR: https://github.com/eclipse-ee4j/yasson/pull/262/files

The new API methods look like....
```java
public interface YassonJsonb extends javax.json.bind.Jsonb {
    /**
     * Reads a {@link JsonStructure} and and converts it into
     * resulting java content tree.
     *
     * @param jsonStructure
     *      {@link JsonStructure} to be used as a source for conversion.
     * @param type
     *      Type of the content tree's root object.
     * @param <T>
     *      Type of the content tree's root object.
     *
     * @return the newly created root object of the java content tree
     *
     * @throws JsonbException
     *     If any unexpected error(s) occur(s) during conversion.
     * @throws NullPointerException
     *      If any of the parameters is {@code null}.
     */
    <T> T fromJsonStructure(JsonStructure jsonStructure, Class<T> type) throws JsonbException;

     /**
     * Reads a {@link JsonStructure} and and converts it into
     * resulting java content tree.
     *
     * @param jsonStructure
     *      {@link JsonStructure} to be used as a source for conversion.
     * @param runtimeType
     *      Runtime type of the content tree's root object.
     * @param <T>
     *      Type of the content tree's root object.
     *
     * @return the newly created root object of the java content tree
     *
     * @throws JsonbException
     *     If any unexpected error(s) occur(s) during deserialization.
     * @throws NullPointerException
     *      If any of the parameters is {@code null}.
     */
    <T> T fromJsonStructure(JsonStructure jsonStructure, Type runtimeType) throws JsonbException;


     /**
     * Serializes the object content tree to a {@link javax.json.JsonStructure}.
     *
     * @param object
     *      The object content tree to be serialized.
     *
     * @return The {@link JsonStructure} serialized from java content tree.
     *
     * @throws JsonbException If any unexpected problem occurs during the
     * serialization.
     * @throws NullPointerException
     *      If any of the parameters is {@code null}.
     *
     * @since JSON Binding 1.0
     */
    JsonStructure toJsonStructure(Object object) throws JsonbException;

     /**
     * Serializes the object content tree to a {@link javax.json.JsonStructure}.
     *
     * @param object
     *      The object content tree to be serialized.
     *
     * @param runtimeType
     *      Runtime type of the content tree's root object.
     *
     * @return The {@link JsonStructure} serialized from java content tree.
     *
     * @throws JsonbException If any unexpected problem occurs during the
     * serialization.
     * @throws NullPointerException
     *      If any of the parameters is {@code null}.
     *
     * @since JSON Binding 1.0
     */
    JsonStructure toJsonStructure (Object object, Type runtimeType) throws JsonbException;

    // <rest of interface unchanged>
}
```

**Comment by michael-schnell** [comment]
Cool. Will someone now take care to move this into the standard API?

**Comment by rmannibucau** [comment]
Why not JsonValue as proposed? Was about to add it to johnzon until i see it was limited to jsonstructure.

**Comment by rmannibucau** [comment]
FYI it is in johnzon with the JsonValue usage - https://issues.apache.org/jira/browse/JOHNZON-220 and https://github.com/apache/johnzon/blob/master/johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/api/experimental/JsonbExtension.java

**Comment by aguibert** [comment]
IIRC we only did `JsonStructure` and `JsonArray` because the other sub-interfaces of `JsonValue` that did not fall under structure/array (`JsonNumber` and `JsonString`) are already better represented as primitives or using methods such as `getNumber`.

@bravehorsie please correct me if I'm remembering this incorrectly

**Comment by rmannibucau** [comment]
Issue is there are numerous library cases where this triggers a type matching which is not needed and kind of defeat these utilities which couldbz then done on top of a parser, no?

**Comment by rmannibucau** [comment]
Hey everyone,

Any hope yasson support <Writer &Consumer<JsonValue>> and <Reader & Supplier<JsonValue>> to have this feature without any signature change?

I just did the change for johnzon >= 1.2.2 which enables to have an optimized runtime when running with johnzon and fallback on a normal stream parsing on other impl but if yasson can do it as well it would make it almost standard - and at least available - without requiring a spec release.

side note: interesting classes: https://gitbox.apache.org/repos/asf?p=johnzon.git;a=blob;f=johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/extension/JsonValueReader.java;h=c4899d380d4c82f72c5e632cca0bc95c032a47a9;hb=d721074adc406c2e5e3112e939379c5989f247a3 and https://gitbox.apache.org/repos/asf?p=johnzon.git;a=blob;f=johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/extension/JsonValueWriter.java;h=fbb9404d65f715abfb54f447164dbf72141e9fd5;hb=d721074adc406c2e5e3112e939379c5989f247a3

Romain

**Comment by scholzi100** [comment]
Is there any workaround for converting between JsonValue <-> Object using yasson yet?
